src/dotnet/SlnFileExtensions.cs

```diff
@@ -87,57 +98,96 @@ public static void AddDefaultBuildConfigurations(this SlnFile slnFile, SlnProjec
-            // stop VS from re-ordering things later on. Since we are keeping the SlnFile class low-level
-            // it shouldn't care about the VS implementation details. That's why we handle this here.
-            AddDefaultSolutionConfigurations(defaultConfigurations, slnFile.SolutionConfigurationsSection);
-            AddDefaultProjectConfigurations(
-                defaultConfigurations,
-                slnFile.ProjectConfigurationsSection.GetOrCreatePropertySet(slnProject.Id));
+            foreach (var config in defaultConfigurations)
+            {
+                configurationsSection[config] = config;
+            }
         }
 
-        private static void AddDefaultSolutionConfigurations(
-            List<string> defaultConfigurations,
-            SlnPropertySet solutionConfigs)
+        private static void MapProjectConfigurations(
+            this SlnFile slnFile,
+            SlnPropertySet projectConfigs,
+            ProjectInstance projectInstance)
         {
-            foreach (var config in defaultConfigurations)
+            // Dictionaries are used here to obtain the original casing of the key
+            var projectConfigurations = new Dictionary<string, string>(StringComparer.CurrentCultureIgnoreCase);
+            var projectPlatforms = new Dictionary<string, string>(StringComparer.CurrentCultureIgnoreCase);
+            string defaultProjectConfiguration = null;
+            string defaultProjectPlatform = null;
+
+            // VS uses the first project configuration/platform as the "default"
+            foreach (var projectConfiguration in projectInstance.GetConfigurations())
             {
-                if (!solutionConfigs.ContainsKey(config))
+                if (defaultProjectConfiguration == null)
                 {
-                    solutionConfigs[config] = config;
+                    defaultProjectConfiguration = projectConfiguration;
                 }
+
+                projectConfigurations[projectConfiguration] = projectConfiguration;
             }
-        }
 
-        private static void AddDefaultProjectConfigurations(
-            List<string> defaultConfigurations,
-            SlnPropertySet projectConfigs)
-        {
-            foreach (var config in defaultConfigurations)
+            foreach (var projectPlatform in projectInstance.GetPlatforms())
             {
-                var activeCfgKey = $"{config}.ActiveCfg";
-                if (!projectConfigs.ContainsKey(activeCfgKey))
+                if (defaultProjectPlatform == null)
                 {
-                    projectConfigs[activeCfgKey] = config;
+                    defaultProjectPlatform = projectPlatform;
                 }
 
-                var build0Key = $"{config}.Build.0";
-                if (!projectConfigs.ContainsKey(build0Key))
+                projectPlatforms[projectPlatform] = projectPlatform;
+            }
+
+            foreach (var key in slnFile.SolutionConfigurationsSection.Keys)
```

Comment:
It feels to me like this whole foreach couple benefit from having its own method to make it easier to understand what it is doing.

Maybe even move the other pieces into submethods as well to make the algorithm clear here.

Reply:
I'll refactor to make the intention clearer.